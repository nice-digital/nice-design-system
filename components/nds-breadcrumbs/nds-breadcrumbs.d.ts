declare module "@nice-digital/nds-breadcrumbs" {
	import React = require("react");

	export interface BreadcrumbProps {
		children: React.ReactNode;
		to?: string;
		elementType?: React.ElementType;
	}

	export const Breadcrumb: React.FC<BreadcrumbProps>;

	export interface BreadcrumbsProps {
		children:
			| React.ReactElement<BreadcrumbProps>[]
			| React.ReactElement<BreadcrumbProps>;
	}

	export const Breadcrumbs: React.FC<BreadcrumbsProps>;
}
