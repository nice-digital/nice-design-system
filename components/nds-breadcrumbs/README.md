# `@nice-digital/nds-breadcrumbs`

> Breadcrumbs component for the NICE Design System

- [`@nice-digital/nds-breadcrumbs`](#nice-digitalnds-breadcrumbs) - [Installation](#installation) - [Usage](#usage) - [React](#react) - [Routing](#routing) - [Props](#props) - [Breadcrumbs](#breadcrumbs) - [children](#children) - [Breadcrumb](#breadcrumb) - [to](#to) - [elementType](#elementtype) - [children](#children-1) - [SCSS](#scss) - [HTML](#html)

## Installation

Install Node, and then:

```sh
npm i @nice-digital/nds-breadcrumbs --save
```

## Usage

### React

Import the `Breadcrumbs` and `Breadcrumb` components from the package and use within JSX:

```jsx
import React from "react";
import { Breadcrumbs, Breadcrumb } from "@nice-digital/nds-breadcrumbs";

<Breadcrumbs>
	<Breadcrumb to="https://www.nice.org.uk">NICE</Breadcrumb>
	<Breadcrumb to="/">CKS</Breadcrumb>
	<Breadcrumb>Topics</Breadcrumb>
</Breadcrumbs>;
```

> Note: The React component automatically imports the SCSS, so there's no need to import the SCSS directly yourself.

#### Routing

To use breadcrumbs with custom routing like [React Router](https://reacttraining.com/react-router/) or [Reach Router](https://reach.tech/router) you can pass in a tag prop. For example to use the `Link` component from [Gatsby](https://www.gatsbyjs.org/docs/gatsby-link/), you can do:

```jsx
import React from "react";
import { Link } from "gatsby";
import { Breadcrumbs, Breadcrumb } from "@nice-digital/nds-breadcrumbs";

<Breadcrumbs>
	<Breadcrumb to="https://www.nice.org.uk">NICE</Breadcrumb>
	<Breadcrumb to="/" tag={Link}>
		CKS
	</Breadcrumb>
	<Breadcrumb>Topics</Breadcrumb>
</Breadcrumbs>;
```

#### Props

##### Breadcrumbs

###### children

- Type: `Breadcrumb`, `Array<Breadcrumb>`

The indvidual breadcrumb items

##### Breadcrumb

###### to

- Type: `string`

The URL the breadcrumb links to

###### elementType

- Type: `null`, `React.ElementType`
- Default: `"a"`

A custom tag type. Defaults to a normal HTMLAnchorElement. Use this prop to use custom routing.

###### children

- Type: `React.ReactNode`

The text of the breadcrumb. Can accept any valid React node, but should be limited to a string to conform with the design system.

### SCSS

If you're not using [React](#react), then import the SCSS directly into your application by:

```scss
@import "~@nice-digital/nds-breadcrumbs/scss/breadcrumbs";
```

### HTML

If you're not using [React](#react), then include the [SCSS as above](#scss) and use the HTML:

```html
<nav aria-label="Breadcrumbs" role="navigation">
	<ol class="breadcrumbs">
		<li class="breadcrumbs__crumb">
			<a href="https://www.nice.org.uk">NICE</a>
		</li>
		<li class="breadcrumbs__crumb">
			<span>CKS</span>
		</li>
	</ol>
</nav>
```
