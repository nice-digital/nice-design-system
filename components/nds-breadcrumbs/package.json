{
  "name": "@nice-digital/nds-breadcrumbs",
  "version": "1.0.25-alpha.0",
  "description": "Breadcrumbs component for the NICE Design System",
  "keywords": [
    "breadcrumbs"
  ],
  "author": "Ian Routledge <ian.routledge@nice.org.uk>",
  "homepage": "https://nhsevidence.github.io/nice-design-system/",
  "license": "MIT",
  "main": "lib/Breadcrumbs.js",
  "module": "es/Breadcrumbs.js",
  "style": "scss/breadcrumbs.scss",
  "types": "nds-breadcrumbs.d.ts",
  "directories": {
    "lib": "lib",
    "test": "__tests__"
  },
  "files": [
    "lib",
    "es",
    "scss",
    "*.d.ts"
  ],
  "publishConfig": {
    "registry": "https://registry.npmjs.org/",
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/nhsevidence/nice-design-system.git"
  },
  "scripts": {
    "test": "echo \"Error: run tests from root\" && exit 1"
  },
  "bugs": {
    "url": "https://github.com/nhsevidence/nice-design-system/issues"
  },
  "dependencies": {
    "@nice-digital/nds-core": "^1.0.23-alpha.0",
    "prop-types": "^15.7.2",
    "react": "^16.11.0"
  },
  "devDependencies": {
    "@babel/cli": "^7.5.0"
  },
  "gitHead": "a4db68f43916ad6c4cb19612e10aae200e80b5a8"
}
