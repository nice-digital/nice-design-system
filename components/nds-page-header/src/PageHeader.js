import React from "react";
import PropTypes from "prop-types";

import "../scss/page-header.scss";

export const PageHeader = props => {
	const { heading, lead, cta, preheading } = props;

	return (
		<div className="page-header">
			<h1 className="page-header__heading">
				{preheading && (
					<span className="page-header__pre-heading">{preheading}</span>
				)}
				{heading}
			</h1>
			{lead && <p className="page-header__lead">{lead}</p>}

			{cta && <p className="page-header__cta">{cta}</p>}
		</div>
	);
};

PageHeader.propTypes = {
	heading: PropTypes.oneOfType([
		PropTypes.arrayOf(PropTypes.node),
		PropTypes.node
	]).isRequired,
	lead: PropTypes.oneOfType([
		PropTypes.arrayOf(PropTypes.node),
		PropTypes.node
	]),
	preheading: PropTypes.oneOfType([
		PropTypes.arrayOf(PropTypes.node),
		PropTypes.node
	]),
	cta: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node])
};
