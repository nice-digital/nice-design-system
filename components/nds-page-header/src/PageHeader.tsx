import React from "react";

import "../scss/page-header.scss";

export interface PageHeaderProps {
	[prop: string]: unknown;
	useAltHeading?: boolean;
	isFullWidth?: boolean;
	preheading?: React.ReactNode;
	breadcrumbs?: React.ReactNode;
	heading: React.ReactNode;
	metadata?: React.ReactNode[];
	lead?: React.ReactNode;
	description?: React.ReactNode;
	cta?: React.ReactNode;
}

export const PageHeader: React.FC<PageHeaderProps> = (
	props: PageHeaderProps
) => {
	const {
		useAltHeading = false,
		isFullWidth = false,
		preheading,
		breadcrumbs,
		heading,
		metadata,
		lead,
		description,
		cta,
		...rest
	} = props;

	// TODO: Refactor this into some sort of conditional component
	// e.g. https://dev.to/dailydevtips1/conditional-wrapping-in-react-46o5
	const PageHeaderContent: React.FC = () => (
		<>
			{breadcrumbs && (
				<div className="page-header__breadcrumbs">{breadcrumbs}</div>
			)}
			<h1
				className={`page-header__heading ${
					useAltHeading ? "page-header__heading--alt" : ""
				}`}
			>
				{preheading && (
					<span className="page-header__pre-heading">{preheading}</span>
				)}
				{heading}
			</h1>

			{lead && <p className="page-header__lead">{lead}</p>}

			{description && (
				<div className="page-header__description">{description}</div>
			)}

			{metadata && (
				<ul className="page-header__metadata">
					{metadata.map((metadatum, i) => (
						<li key={i}>{metadatum}</li>
					))}
				</ul>
			)}

			{cta && <p className="page-header__cta">{cta}</p>}
		</>
	);

	return (
		<div
			className={`page-header ${isFullWidth ? "page-header--full-width" : ""}`}
			data-component="page-header"
			{...rest}
		>
			{isFullWidth ? (
				<div className="container">
					<PageHeaderContent />
				</div>
			) : (
				<PageHeaderContent />
			)}
		</div>
	);
};
