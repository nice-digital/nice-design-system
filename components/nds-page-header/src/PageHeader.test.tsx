"use strict";
import React from "react";
import { render } from "@testing-library/react";
import { PageHeader, PageHeaderProps } from "./PageHeader";
import { Breadcrumb, Breadcrumbs } from "@nice-digital/nds-breadcrumbs";

const breadcrumbs: React.ReactNode = (
	<Breadcrumbs>
		<Breadcrumb to="https://www.nice.org.uk/">Home</Breadcrumb>
		<Breadcrumb to="https://www.nice.org.uk/guidance">NICE guidance</Breadcrumb>
		<Breadcrumb>Published</Breadcrumb>
	</Breadcrumbs>
);

const secondSection: React.ReactNode = (
	<aside>
		<h3>I am a second section</h3>
		<ol>
			<li>One</li>
			<li>Two</li>
			<li>Three</li>
		</ol>
	</aside>
);

const props: PageHeaderProps = {
	heading: "Products",
	cta: "See more below",
	preheading: "Systemic lupus erythematosus",
	lead: "A list of all our products on systemic lupus erythematosus",
	description: "A description of whatever is on the page",
	breadcrumbs,
	variant: "fullWidthDark",
	secondSection
};

describe("PageHeader", () => {
	it("should match the snapshot with all props supplied", () => {
		const wrapper = render(<PageHeader {...props} />);
		expect(wrapper).toMatchSnapshot();
	});

	it("should match the snapshot with only required heading prop supplied", () => {
		const wrapper = render(<PageHeader heading="This is our heading" />);
		expect(wrapper).toMatchSnapshot();
	});

	it("should allow a component to be passed as a prop", () => {
		const button = <button onClick={() => false}>Click Me</button>;
		const wrapper = render(
			<PageHeader heading="This is our heading" cta={button} />
		);
		expect(wrapper).toMatchSnapshot();
	});

	it("should pass additional props to the container", () => {
		const { container } = render(
			<PageHeader heading="This is our heading" data-track={false} />
		);
		expect(container.querySelector("div")?.getAttribute("data-track")).toBe(
			"false"
		);
	});

	it("should merge classes with the container", () => {
		const { container } = render(
			<PageHeader heading="This is our heading" className="mt--0" />
		);
		expect(container.querySelector("div")).toHaveClass("mt--0");
	});
});
