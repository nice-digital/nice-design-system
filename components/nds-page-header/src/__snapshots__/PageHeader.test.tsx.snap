// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageHeader should allow a component to be passed as a prop 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="page-header "
        data-component="page-header"
      >
        <h1
          class="page-header__heading "
        >
          This is our heading
        </h1>
        <p
          class="page-header__cta"
        >
          <button>
            Click Me
          </button>
        </p>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="page-header "
      data-component="page-header"
    >
      <h1
        class="page-header__heading "
      >
        This is our heading
      </h1>
      <p
        class="page-header__cta"
      >
        <button>
          Click Me
        </button>
      </p>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`PageHeader should match the snapshot with all props supplied 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="page-header page-header--full-width-dark"
        data-component="page-header"
      >
        <div
          class="container"
        >
          <div
            class="page-header__breadcrumbs"
          >
            <nav
              aria-label="Breadcrumbs"
              data-component="breadcrumbs"
              role="navigation"
            >
              <ol
                class="breadcrumbs"
              >
                <li
                  class="breadcrumbs__crumb"
                >
                  <a
                    href="https://www.nice.org.uk/"
                  >
                    Home
                  </a>
                </li>
                <li
                  class="breadcrumbs__crumb"
                >
                  <a
                    href="https://www.nice.org.uk/guidance"
                  >
                    NICE guidance
                  </a>
                </li>
                <li
                  class="breadcrumbs__crumb"
                >
                  <span>
                    Published
                  </span>
                </li>
              </ol>
            </nav>
            <script
              type="application/ld+json"
            >
              {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "https://www.nice.org.uk/",
        "name": "Home"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@id": "https://www.nice.org.uk/guidance",
        "name": "NICE guidance"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "name": "Published"
      }
    }
  ]
}
            </script>
          </div>
          <h1
            class="page-header__heading "
          >
            <span
              class="page-header__pre-heading"
            >
              Systemic lupus erythematosus
            </span>
            Products
          </h1>
          <p
            class="page-header__lead"
          >
            A list of all our products on systemic lupus erythematosus
          </p>
          <div
            class="page-header__description"
          >
            A description of whatever is on the page
          </div>
          <p
            class="page-header__cta"
          >
            See more below
          </p>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="page-header page-header--full-width-dark"
      data-component="page-header"
    >
      <div
        class="container"
      >
        <div
          class="page-header__breadcrumbs"
        >
          <nav
            aria-label="Breadcrumbs"
            data-component="breadcrumbs"
            role="navigation"
          >
            <ol
              class="breadcrumbs"
            >
              <li
                class="breadcrumbs__crumb"
              >
                <a
                  href="https://www.nice.org.uk/"
                >
                  Home
                </a>
              </li>
              <li
                class="breadcrumbs__crumb"
              >
                <a
                  href="https://www.nice.org.uk/guidance"
                >
                  NICE guidance
                </a>
              </li>
              <li
                class="breadcrumbs__crumb"
              >
                <span>
                  Published
                </span>
              </li>
            </ol>
          </nav>
          <script
            type="application/ld+json"
          >
            {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "https://www.nice.org.uk/",
        "name": "Home"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@id": "https://www.nice.org.uk/guidance",
        "name": "NICE guidance"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "name": "Published"
      }
    }
  ]
}
          </script>
        </div>
        <h1
          class="page-header__heading "
        >
          <span
            class="page-header__pre-heading"
          >
            Systemic lupus erythematosus
          </span>
          Products
        </h1>
        <p
          class="page-header__lead"
        >
          A list of all our products on systemic lupus erythematosus
        </p>
        <div
          class="page-header__description"
        >
          A description of whatever is on the page
        </div>
        <p
          class="page-header__cta"
        >
          See more below
        </p>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`PageHeader should match the snapshot with only required heading prop supplied 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="page-header "
        data-component="page-header"
      >
        <h1
          class="page-header__heading "
        >
          This is our heading
        </h1>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="page-header "
      data-component="page-header"
    >
      <h1
        class="page-header__heading "
      >
        This is our heading
      </h1>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
